systemd: build cryptsetup

vigh-m opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes https://github.com/bottlerocket-os/bottlerocket-core-kit/issues/536

**Description of changes:**
- Build `systemd-cryptsetup` for `systemd-257`
- Also update `systemd-257` to the latest upstream release.
   - This is required since upstream has added support to build `systemd` without `openssl/ui.h` which is a requirement to build `systemd-cryptsetup`

**Testing done:**

All testing was done on an aws-k8s-1.34(-nvidia) variant

- No errors in the journalctl logs:
   ```
   Oct 10 17:29:29 localhost kernel: Run /sbin/init as init process
   Oct 10 17:29:29 localhost kernel:   with arguments:
   Oct 10 17:29:29 localhost kernel:     /sbin/init
   Oct 10 17:29:29 localhost kernel:     systemd.log_target=journal-or-kmsg
   Oct 10 17:29:29 localhost kernel:     systemd.log_color=0
   Oct 10 17:29:29 localhost kernel:     systemd.show_status=true
   Oct 10 17:29:29 localhost kernel:   with environment:
   Oct 10 17:29:29 localhost kernel:     HOME=/
   Oct 10 17:29:29 localhost kernel:     TERM=linux
   Oct 10 17:29:29 localhost kernel:     SYSTEMD_DEFAULT_MOUNT_RATE_LIMIT_BURST=25
   Oct 10 17:29:29 localhost kernel:     SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1
   Oct 10 17:29:29 localhost kernel:     BOOT_IMAGE=(hd0,gpt3)/vmlinuz
   Oct 10 17:29:29 localhost kernel:     selinux=1
   Oct 10 17:29:29 localhost kernel:     enforcing=1
   .
   .
   .
   Oct 10 17:29:29 localhost systemd[1]: systemd 257.9 running in system mode (-PAM -AUDIT +SELINUX -APPARMOR -IMA +IPE -SMACK +SECCOMP -GCRYPT -GNUTLS +OPENSSL +ACL +BLKID -CURL -ELFUTILS -FIDO2 -IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBCRYPTSETUP_PLUGINS +LIBFDISK -PCRE2 -PWQUALITY -P11KIT -QRENCODE +TPM2 -BZIP2 -LZ4 -XZ -ZLIB -ZSTD -BPF_FRAMEWORK -BTF -XKBCOMMON -UTMP -SYSVINIT -LIBARCHIVE)
   ```


- Tested a few systemd commands without error
   ```
   bash-5.1# systemd-analyze
   Startup finished in 798ms (firmware) + 644ms (loader) + 2.376s (kernel) + 13.863s (userspace) = 17.682s
   multi-user.target reached after 13.824s in userspace.

   bash-5.1# systemctl restart kubelet
   bash-5.1# systemctl status kubelet
   ● kubelet.service - Kubelet
        Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service; enabled; preset: enabled)
       Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
                └─00-aws-config.conf
                /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service.d
                └─dockershim-symlink.conf
                /etc/systemd/system/kubelet.service.d
                └─exec-start.conf
                /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service.d
                └─make-kubelet-dirs.conf, prestart-load-pause-ctr.conf
       Active: active (running) since Fri 2025-10-10 17:35:03 UTC; 1s ago
    Invocation: 8b9155cd79764e9193afb39a475e60b2
          Docs: https://github.com/kubernetes/kubernetes
       Process: 4957 ExecStartPre=/sbin/iptables -P FORWARD ACCEPT (code=exited, status=0/SUCCESS)
       Process: 4958 ExecStartPre=/bin/ln -sf ./containerd/containerd.sock /run/dockershim.sock (code=exited, status=0/SUCCESS)
       Process: 4961 ExecStartPre=/usr/bin/mkdir -p /var/lib/kubelet/providers/secrets-store /var/lib/kubelet/node-feature-discovery/features.d (code=exited, status=0/SUCCESS)
       Process: 4963 ExecStartPre=/usr/bin/ctr --namespace=k8s.io image import --all-platforms /usr/libexec/kubernetes/kubernetes-pause.tar (code=exited, status=0/SUCCESS)
       Process: 4969 ExecStartPre=/usr/bin/ctr --namespace=k8s.io image label localhost/kubernetes/pause:0.1.0 io.cri-containerd.pinned=pinned (code=exited, status=0/SUCCESS)
      Main PID: 4974 (kubelet)
         Tasks: 11 (limit: 9313)
        Memory: 30.5M (peak: 38M)
           CPU: 381ms
        CGroup: /runtime.slice/kubelet.service
                └─4974 /usr/bin/kubelet --cloud-provider external --kubeconfig /etc/kubernetes/kubelet/kubeconfig --config /etc/kubernetes/kubelet/config --container-runtime-endpoint=unix:///run/containerd/containerd.sock --containerd=/run/containerd/containerd.sock --root-dir /var/lib/kubelet --cert-dir /var/lib/kubel…

   ```


- No unit files showing any errors
   ```
   bash-5.1# systemctl | grep -e "error" -e "ERROR" -e "Error" -e "failed" -e "FAILED" -e  "Failed"
   bash-5.1#
   ```

- The instance joins the cluster as expected

- Basic workloads are launched on the node correctly

- Tested neuron instances:
   ```
   bash-5.1# systemctl status modprobe@neuron
   ● modprobe@neuron.service - Load Kernel Module neuron
        Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/modprobe@.service; static)
       Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
                └─00-aws-config.conf
                /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/modprobe@.service.d
                └─10-remain-after-exit.conf
                /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/modprobe@neuron.service.d
                └─neuron.conf
        Active: active (exited) since Fri 2025-10-10 18:30:49 UTC; 4min 33s ago
    Invocation: dd5a2c8f3c434733be23f8067fc24207
          Docs: man:modprobe(8)
      Main PID: 1306 (code=exited, status=0/SUCCESS)
      Mem peak: 2.8M
           CPU: 39ms

   Oct 10 18:30:49 localhost systemd[1]: Finished Load Kernel Module neuron.
   Notice: journal has been rotated since unit was started, output may be incomplete.
   ```

   ```
   % kubectl get nodes "-o=custom-columns=NAME:.metadata.name,NeuronCore:.status.allocatable.aws\.amazon\.com/neuroncore"
   NAME                                          NeuronCore
   ip-192-168-70-48.us-west-2.compute.internal   2
   ```


- Tested Neuron Workloads:
   ```
   === RUN   TestNeuronNodes/multi-node
       neuron_test.go:157: Skipping feature "multi-node": name not matched
   --- PASS: TestNeuronNodes (360.09s)
       --- PASS: TestNeuronNodes/single-node (360.09s)
           --- PASS: TestNeuronNodes/single-node/Single_node_test_Job_succeeds (360.02s)
      --- SKIP: TestNeuronNodes/multi-node (0.00s)
   PASS
   ok      github.com/aws/aws-k8s-tester/test/cases/neuron 431.977s
   ```


- Nvidia instances working as expected:
   ```
   bash-5.1# nvidia-smi
   Fri Oct 10 21:23:46 2025
   +-----------------------------------------------------------------------------------------+
   | NVIDIA-SMI 580.95.05              Driver Version: 580.95.05      CUDA Version: 13.0     |
   +-----------------------------------------+------------------------+----------------------+
   | GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
   | Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
   |                                         |                        |               MIG M. |
   |=========================================+========================+======================|
   |   0  Tesla T4                       On  |   00000000:00:1E.0 Off |                    0 |
   | N/A   31C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
   |                                         |                        |                  N/A |
   +-----------------------------------------+------------------------+----------------------+
   
   +-----------------------------------------------------------------------------------------+
   | Processes:                                                                              |
   |  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
   |        ID   ID                                                               Usage      |
   |=========================================================================================|
   |  No running processes found                                                             |
   +-----------------------------------------------------------------------------------------+
   ```

- Basic Nvidia workload tests worked as expected.


- Ran basic scale testing by launching ~100 g4dn.xlarge, m5.large, c6i.large, r7a.large, 

- [x] Soak testing with an AMI

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

vigh-m committed:
systemd: update systemd-257 to v257.9

vigh-m committed:
systemd: remove openssl/ui support

This extends the upstream patch to allow building systemd with openssl
drop-ins that don't have UI support

vigh-m committed:
systemd: fix type mismatch between openssl and aws-lc

openssl and aws-lc (and boringssl) have diverged wrt the return type of
ERR_get_error() and `long unsigned int` has been patched to be
`uint32_t` instead

vigh-m committed:
systemd: remove unsupported cipher

vigh-m committed:
systemd: disable systemd-sb-sign

sb-sign has a dependency on PKCS7 which is not provided by aws-lc.
Adding a new meson option to prevent it from being built

vigh-m committed:
systemd: stub out an unsupported method in bootctl

Stub out install_secure_boot_auto_enroll since it depends on PKCS7. Instead
default to the EOPNOTSUPP condition with a debug log

vigh-m committed:
systemd: build cryptsetup

vigh-m commented:
⬆️ split the changes into multiple commits

bcressey commented on packages/systemd-257/9012-openssl-util-build-without-ui.patch:
```diff
@@ -0,0 +1,53 @@
+ #  ifndef OPENSSL_NO_UI_CONSOLE
+ #    include <openssl/ui.h>
++#  else
++#    include <openssl/hmac.h>
```
This should have its own guard:
```
#  ifndef OPENSSL_HMAC_H
#    include <openssl/hmac.h>
#  endif
```

bcressey commented on packages/systemd-257/9013-fix-openssl-aws-lc-divergence-in-data-types.patch:
```diff
@@ -0,0 +1,59 @@
+         if (r < 0)
+                 return log_debug_errno(SYNTHETIC_ERRNO(EIO),
+-                                       "Signature verification failed: 0x%lx", ERR_get_error());
++                                       "Signature verification failed: 0x%u", ERR_get_error());
```
If this is really a uint32_t, you should use the matching printf specifier:
```
#include <inttypes.h>
...
"Signature verification failed: 0x"PRIx32, ERR_get_error()

vigh-m replied:
Yeah, [aws-lc](https://github.com/aws/aws-lc/blob/0fed4dbc456406bdaf7158fd55d05a52d9998ce4/include/openssl/err.h#L185) and [open-ssl](https://github.com/openssl/openssl/blob/5aaf9746f649d15b3ad006acdc6a958819acc536/include/openssl/err.h.in#L406) have different implementations for this.

bcressey commented on packages/systemd-257/9012-openssl-util-build-without-ui.patch:
```diff
@@ -0,0 +1,53 @@
+ DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ASN1_TYPE*, ASN1_TYPE_free, NULL);
+ DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ASN1_STRING*, ASN1_STRING_free, NULL);
+ 
++#  ifndef OPENSSL_NO_UI_CONSOLE
```
nit: spacing is off here

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -3,8 +3,11 @@
 %global package_priority_epoch 0
 
+# Disable OpenSSL UI since aws-lc does not support it.
+%global _cross_cflags %{_cross_cflags} -DOPENSSL_NO_UI_CONSOLE=1
```
I'd slightly prefer either:
1. patching the pkgconfig file in `libcrypto` to include this by default
2. patching `meson.build` to include it

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -57,6 +60,24 @@ Patch9010: 9010-meson-replace-openssl-dependency-with-libcrypto.patch
+
+# Disable sb-sign since that has a dependency on PKCS7 which is not provided
+# by aws-lc
+ Patch9015: 9015-disable-sb-sign.patch
```
```suggestion
# Disable sb-sign since that has a dependency on PKCS7 which is not provided
# by aws-lc
Patch9015: 9015-disable-sb-sign.patch
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -785,3 +813,45 @@ install -p -m 0644 %{S:2} %{buildroot}%{_cross_bootconfigdir}/21-cgroup-enable-l
+%{_cross_libdir}/pcrlock.d/850-sysinit.pcrlock
+%{_cross_libdir}/pcrlock.d/900-ready.pcrlock
+%{_cross_libdir}/pcrlock.d/950-shutdown.pcrlock
+%{_cross_libdir}/pcrlock.d/990-final.pcrlock
```
I doubt that these specific pcrlock policy files will be useful, though the mechanism is elegant.

I'd like to get `systemd-pcrextend` included; that's what would perform some of the measurements that these pcrlock files anticipate.

vigh-m replied:
Ack. I've `%exclude(d)` them in their own section for now.  `systemd-pcrextend` requires  [BOOTLOADER](https://github.com/systemd/systemd/blob/48aec295a856476aa24b3a6f8c18caa77c5fbe37/src/pcrextend/meson.build#L8) to be enabled which I can explore outside of this

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -785,3 +813,45 @@ install -p -m 0644 %{S:2} %{buildroot}%{_cross_bootconfigdir}/21-cgroup-enable-l
+%{_cross_libdir}/pcrlock.d/990-final.pcrlock
+%{_cross_libdir}/systemd/system-generators/systemd-cryptsetup-generator
+%{_cross_libdir}/systemd/system-generators/systemd-integritysetup-generator
+%{_cross_libdir}/systemd/system-generators/systemd-veritysetup-generator
```
```suggestion
%{_cross_systemdgeneratordir}/systemd-cryptsetup-generator
%{_cross_systemdgeneratordir}/systemd-integritysetup-generator
%{_cross_systemdgeneratordir}/systemd-veritysetup-generator
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -785,3 +813,45 @@ install -p -m 0644 %{S:2} %{buildroot}%{_cross_bootconfigdir}/21-cgroup-enable-l
+%{_cross_libdir}/systemd/system-generators/systemd-integritysetup-generator
+%{_cross_libdir}/systemd/system-generators/systemd-veritysetup-generator
+%{_cross_libdir}/systemd/system/cryptsetup-pre.target
+%{_cross_libdir}/systemd/system/cryptsetup.target
```
Use `%{_cross_unitdir}` instead of `%{_cross_libdir}/systemd/system`.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -785,3 +813,45 @@ install -p -m 0644 %{S:2} %{buildroot}%{_cross_bootconfigdir}/21-cgroup-enable-l
+%{_cross_libdir}/systemd/system/cryptsetup-pre.target
+%{_cross_libdir}/systemd/system/cryptsetup.target
+%{_cross_libdir}/systemd/system/initrd-root-device.target.wants/remote-cryptsetup.target
+%{_cross_libdir}/systemd/system/initrd-root-device.target.wants/remote-veritysetup.target
```
omit: we don't have an initrd
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -785,3 +813,45 @@ install -p -m 0644 %{S:2} %{buildroot}%{_cross_bootconfigdir}/21-cgroup-enable-l
+%{_cross_libdir}/systemd/system/integritysetup-pre.target
+%{_cross_libdir}/systemd/system/integritysetup.target
+%{_cross_libdir}/systemd/system/remote-cryptsetup.target
+%{_cross_libdir}/systemd/system/remote-veritysetup.target
```
Add these to the "Exclude remote filesystem targets" section above.

bcressey reviewed (commented):
Can you run this again now that cryptsetup is enabled and fix any spurious differences?
```
git diff --no-index packages/systemd-25{2,7}/systemd*.spec
```

vigh-m committed:
systemd: add meson option to set openssl ui cargs

Add a patch to extend meson options to set OPENSSL_NO_UI_CONSOLE=1
during build. The option can be controlled with the
CONFIGURE_OPTS in the specfile

vigh-m commented:
⬆️ Addressed comments

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -794,6 +831,30 @@ install -p -m 0644 %{S:2} %{buildroot}%{_cross_bootconfigdir}/21-cgroup-enable-l
+%{_cross_libdir}/systemd/system/system-systemd\x2dcryptsetup.slice
+%{_cross_libdir}/systemd/system/system-systemd\x2dveritysetup.slice
+%{_cross_libdir}/systemd/system/veritysetup-pre.target
+%{_cross_libdir}/systemd/system/veritysetup.target
```
replace `%{_cross_libdir}/systemd/system` with `%{_cross_unitdir}`.

vigh-m commented:
⬆️ replaced `%{_cross_libdir}/systemd/system` with `%{_cross_unitdir}`. Also moved things around a little to reduce the diff between systemd252 and systemd257 specs
